lib/testing/src/equality/equality.dart

```diff
@@ -0,0 +1,229 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+part of quiver.testing.equality;
+
+
+/**
+ * Matcher for == and hashCode methods of a class.
+ *
+ * To use, invoke areEqualityGroups with a list of equality groups where each
+ * group contains objects that are supposed to be equal to each other, and
+ * objects of different groups are expected to be unequal. For example:
+ *
+ * expect({
+ *      'hello': ["hello", "h" + "ello"],
+ *      'world': ["world", "wor" + "ld"],
+ *      'three': [2, 1 + 1]
+ *     }, areEqualityGroups);
+ *
+ * This tests:
+ *
+ * * comparing each object against itself returns true
+ * * comparing each object against an instance of an incompatible class
+ *     returns false
+ * * comparing each pair of objects within the same equality group returns
+ *     true
+ * * comparing each pair of objects from different equality groups returns
+ *     false
+ * * the hash codes of any two equal objects are equal
+ *
+ *
+ * When a test fails, the error message labels the objects involved in
+ * the failed comparison as follows:
+ *
+ *      "`[group i, item j]`" refers to the jth item in the ith equality group,
+ *       where both equality groups and the items within equality groups are
+ *       numbered starting from 1.  When either a constructor argument or an
+ *       equal object is provided, that becomes group 1.
+ *
+ */
+const REPETITIONS = 3;
+const NO_MISMATCH_FOUND = 'therewasnomismatchfoundhere';
+const Matcher areEqualityGroups = const _EqualityGroupMatcher();
+
+class _EqualityGroupMatcher extends Matcher {
+  static const FAILURE_REASON = 'failureReason';
+  const _EqualityGroupMatcher();
+
+  @override
+  Description describe(Description description) =>
+      description.add('to be equality groups');
+
+  @override
+  bool matches(Map<String, List<Object>> item, Map matchState) {
+    try {
+      _verifyEqualityGroups(item, matchState);
+      return true;
+    } on MatchError catch (e) {
+      matchState[FAILURE_REASON] = e.toString();
+      return false;
+    }
+  }
+
+  Description describeMismatch(item, Description mismatchDescription,
+      Map matchState, bool verbose) => mismatchDescription
+        .add(" ${matchState[FAILURE_REASON]}");
+
+
+  void _verifyEqualityGroups(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    if (equalityGroups == null) {
+      throw new MatchError('Equality Group must not be null');
+    }
+    var equalityGroupsCopy = {};
+    equalityGroups.keys.forEach((String groupName) {
+      if (groupName == null) {
+        throw new MatchError('Group name must not be null');
+      }
+      var group = equalityGroups[groupName];
+      if (group == null) {
+        throw new MatchError('Group must not be null');
+      }
+      equalityGroupsCopy[groupName] = new List.from(group);
```

Comment:
Consider either using `List` everywhere, or use `List<Object>.from(group)` here for consistency.